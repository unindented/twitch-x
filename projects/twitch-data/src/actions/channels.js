import {getAction} from '../api'

export function loadSearchChannels ({query, limit}) {
  const action = {
    type: 'LOAD_SEARCH_CHANNELS',
    payload: query
  }
  return getAction(action, '/search/channels', {query: {query: encodeURIComponent(query), limit}})
}
