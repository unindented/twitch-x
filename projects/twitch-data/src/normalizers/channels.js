import {normalize, schema} from 'normalizr'
import {valueOrNull} from './utils'

const channel = new schema.Entity('channels', {}, {
  idAttribute: 'name',
  processStrategy: (channel) => {
    return {
      id: valueOrNull(channel.name),
      status: valueOrNull(channel.status),
      name: valueOrNull(channel.display_name),
      game: valueOrNull(channel.game),
      image: valueOrNull(channel.logo)
    }
  }
})

const root = [channel]

export function normalizeSearchChannels (payload) {
  return normalize(payload.channels, root)
}
