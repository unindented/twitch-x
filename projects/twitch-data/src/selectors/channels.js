import {createSelector} from 'reselect'

export const getSearchChannels = createSelector(
  (state) => state.data.channels.search,
  (state) => state.data.channels.byId,
  (searchChannelIds, channelsById) => searchChannelIds.map((id) => channelsById[id])
)
