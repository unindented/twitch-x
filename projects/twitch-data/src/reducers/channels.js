import update from 'immutability-helper'
import {normalizeSearchChannels} from '../normalizers/channels'

const initialState = {
  byId: {},
  search: []
}

export default (state = initialState, action = {}) => {
  switch (action.type) {
    case 'LOAD_SEARCH_CHANNELS_SUCCESS': {
      const {entities, result} = normalizeSearchChannels(action.payload.response)
      return update(state, {
        byId: {$merge: entities.channels},
        search: {$set: result}
      })
    }
  }

  return state
}
