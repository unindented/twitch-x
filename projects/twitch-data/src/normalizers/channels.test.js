import {normalizeSearchChannels} from './channels'

const fixture = require('../__fixtures__/searchChannels.json')

describe('normalizers/channels', () => {
  describe('.normalizeSearchChannels', () => {
    it('normalizes the payload', () => {
      const result = normalizeSearchChannels(fixture)
      expect(result).toMatchSnapshot()
    })
  })
})
