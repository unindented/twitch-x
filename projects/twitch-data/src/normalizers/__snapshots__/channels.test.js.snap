// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizers/channels .normalizeSearchChannels normalizes the payload 1`] = `
Object {
  "entities": Object {
    "channels": Object {
      "day9": Object {
        "game": null,
        "id": "day9",
        "image": null,
        "name": "day9",
        "status": null,
      },
      "day9daily": Object {
        "game": null,
        "id": "day9daily",
        "image": "https://static-cdn.jtvnw.net/jtv_user_pictures/day9daily-profile_image-14e040a6b9a079be-300x300.jpeg",
        "name": "day9daily",
        "status": "LIVE from my iPhone 4!",
      },
      "day9tv": Object {
        "game": "Dota 2",
        "id": "day9tv",
        "image": "https://static-cdn.jtvnw.net/jtv_user_pictures/day9tv-profile_image-2226c4b27aaf487a-300x300.png",
        "name": "Day9tv",
        "status": "10v10 DayKnight Festival!  Play Games w/ us!  discord.gg/day9tv",
      },
      "deaths_day98": Object {
        "game": "Battlefield Hardline",
        "id": "deaths_day98",
        "image": null,
        "name": "Deaths_day98",
        "status": "BFH with friends",
      },
    },
  },
  "result": Array [
    "day9tv",
    "day9",
    "day9daily",
    "deaths_day98",
  ],
}
`;
